doc/adr/0002-nats-typed-messages.md

```diff
@@ -0,0 +1,150 @@
+those to other cloud systems using the `jsm.go` package that supports converting `events` into Cloud Event format.
+
+### Message Types
+
+There is no standard way to indicate the type of a specific message. We looked at a lot of prior art from CNCF
+projects, public clouds and more but found very little commonality. The nearest standard is the Uniform Resource Name
+which still leaves most of the details up to the project and does not conventionally support versioning.
+
+We chose a message type like `io.nats.jetstream.api.v1.consumer_delete_response`, `io.nats.server.advisory.v1.client_connect`
+or `io.nats.unknown_message`.
+
+`io.nats.unknown_message` is a special type returned for anything without valid type hints. In go that implies `map[string]interface{}`.
+
+The structure is as follows: io.nats.`<project>`.`<subsystem or type>`.v`<version>`.`<message name>`
+
+#### Project
+
+The project is the overall project this belongs to and should be short but descriptive, today we have 2 - `server` and `jetstream` - as we continue to 
+build systems around Stream Processing and more we'd add more of these types. I anticipate for example adding
+a few to Surveyor for publishing significant lifecycle events.
+
+|Project|Description|
+|-------|-----------|
+|`server`|The core NATS Server excluding JetStream related messages|
+|`jetstream`|Any JetStream related message
+
+#### Subsystem or Type
+
+This is a bit undefined, examples in use now are `api`, `advisory`, `metric`. 
+
+|Type|Description|
+|----|-----------|
+|`api`|Typically these are `messages` used in synchronous request response APIs|
+|`advisory`|These are `events` that describe a significant event that happened like a client connecting or disconnecting|
+|`metric`|These are `events` that relate to monitoring - how long did it take a message to be acknowledged|
+
+#### Versioning
+
+The ideal outcome is that we never need to version any message and maintain future compatibility. 
+
+We think we can do that with the JetStream API. Monitoring, observability and black box management is emerging, and we 
+know less about how that will look in the long run, so we think we will need to version those.
+
+So all message types includes a single digit version.
+
+#### Message Name
+
+Just a string identifying what this message is about - `client_connect`, `client_disconnect`, `api_audit` etc.
+
+## Examples
+
+### Messages
+
+At minimum a typed message must include a `type` string:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_configuration"
+}
+```
+
+Rest of the document is up to the specific use case
+
+### Advisories
+
+Advisories must include additional fields:
+
+```json
+{
+  "type": "io.nats.jetstream.advisory.v1.api_audit",
+  "id": "uafvZ1UEDIW5FZV6kvLgWA",
+  "timestamp": "2020-04-23T16:51:18.516363Z"
+}
+```
+
+ * `timestamp` - RFC 3339 format in UTC timezone, with sub-second precision added if present
+ * `id` - Any sufficiently unique ID such as those produced by `nuid`
+
```

Comment:
Call out a minimum error?

Reply:
yeah good call, will add

Reply:
actually, I'll mention them but this is not a feature of either message or event.  One can add them if you want and JetStream API does so, but it's not a core part of either message or event.  

I'll mention them and point to the 0001 ADR though

Reply:
Added in https://github.com/nats-io/nats-server/pull/1374/commits/53b3b914c147fe53b60bf5ab66946cdbf1940d26